codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -0,0 +1,686 @@
+          SmithyNameResolver.getGoModuleNameForSmithyNamespace(
+            resourceOrService.toShapeId().getNamespace()
+          ),
+          SmithyNameResolver.shapeNamespace(serviceShape)
+        );
+        namespace =
+          SmithyNameResolver.shapeNamespace(resourceOrService).concat(".");
+      }
+      if (!this.isOptional) {
+        return "return %1$s{%2$s}".formatted(
+            namespace.concat(
+              context.symbolProvider().toSymbol(serviceShape).getName()
+            ),
+            dataSource
+          );
+      }
+      return """
+      return func () *%s {
+          if %s == nil {
+              return nil;
+          }
+          return &%s{%s.(*%s)}
+      }()""".formatted(
+          namespace.concat(
+            context.symbolProvider().toSymbol(serviceShape).getName()
+          ),
+          dataSource,
+          namespace.concat(
+            context.symbolProvider().toSymbol(serviceShape).getName()
+          ),
+          dataSource,
+          DafnyNameResolver.getDafnyClient(serviceShape.toShapeId().getName())
+        );
+    }
+  }
+
+  @Override
+  protected String getDefault(final Shape shape) {
+    throw new CodegenException(
+      String.format(
+        "Unsupported conversion of %s to %s using the %s protocol",
+        shape,
+        shape.getType(),
+        context.protocolGenerator().getName()
+      )
+    );
+  }
+
+  @Override
+  public String blobShape(final BlobShape shape) {
+    writer.addImportFromModule(DAFNY_RUNTIME_GO_LIBRARY_MODULE, "dafny");
+    return """
+    return func () []byte {
+    var b []byte
+    if %s == nil {
+        return nil
+    }
+    for i := dafny.Iterate(%s) ; ; {
+        val, ok := i()
+        if !ok {
+            return b
+        } else {
+            b = append(b, val.(byte))
+        }
+    }
+    }()""".formatted(dataSource, dataSource);
+  }
+
+  @Override
+  public String structureShape(final StructureShape shape) {
+    if (shape.hasTrait(ReferenceTrait.class)) {
+      return referenceStructureShape(shape);
+    }
+    final var builder = new StringBuilder();
+    writer.addImportFromModule(
+      SmithyNameResolver.getGoModuleNameForSmithyNamespace(
+        shape.toShapeId().getNamespace()
+      ),
+      DafnyNameResolver.dafnyTypesNamespace(shape)
+    );
+    final String referenceType = (this.isOptional) ? "&" : "";
+    builder.append(
+      "return %1$s%2$s{".formatted(
+          referenceType,
+          SmithyNameResolver
+            .smithyTypesNamespace(shape)
+            .concat(".")
+            .concat(shape.getId().getName())
+        )
+    );
+    for (final var memberShapeEntry : shape.getAllMembers().entrySet()) {
+      final var memberName = memberShapeEntry.getKey();
+      final var memberShape = memberShapeEntry.getValue();
+      final var targetShape = context
+        .model()
+        .expectShape(memberShape.getTarget());
+      //TODO: Is it ever possible for structure to be nil?
```

Comment:
One approach to this lets you say the answer is "no":
If the structure is required, it will never be nil
If it is optional, you can check if it is nil before going into the structure converter:

```
nativeOutput = (input.isNotNil ? structureConverter(input) : nil)
```

and only go into the structure converter if it is not nil

But you might need to make a code change to get the optional checking in

Reply:
It is more nuanced than that. A member shape can be Optional but still not be a pointer. Nil check can only operate on pointer types and not on value types. A structure can be value type or pointer type in Go SDK depending on some set of rules - and that set of rules is not very well defined for Dafny vs SDK. Hence the TODO. We added this so that if something fails tomorrow for Structure being nil - we know where to look.